lib/remote/jsonrpcconnection.hpp

```diff
@@ -63,8 +63,10 @@ class JsonRpcConnection final : public Object
 	static double GetWorkQueueRate();
 
 	static void SendCertificateRequest(const JsonRpcConnection::Ptr& aclient, const intrusive_ptr<MessageOrigin>& origin, const String& path);
+	static void SendCertificateRequest(JsonRpcConnection* aclient, const intrusive_ptr<MessageOrigin>& origin, const String& path);
```

Comment:
Otherwise the JSON RPC connection holds a shared pointer to the timer which holds a shared pointer to the lambda which holds a shared (not regular!) pointer to the JSON RPC connection.

Reply:
Okay, so to break cyclic references, i.e. to basically get the same effect as `JsonRpcConnection *c = aclient.get();` and then capturing that in the lambda instead?

Maybe another reason for https://github.com/Icinga/icinga2/pull/9179#discussion_r839534223? With that you could just start the timer in a member function, capture `this` and things would be fine.

Reply:
> capture `this`

... and then? Would this avoid the overload?

Reply:
That would be two different functions, so yes.

Reply:
No. Imagine the connection is being destroyed, so the timer is being destroyed, so the timer is being waited for because it's running while it calls the (only) shared ptr overload and causing the refcount 0->1 and then 1->0->delete.